src/simularium/VisDataParse.ts

```diff
@@ -1,119 +1,109 @@
+                `Your data is malformed, there are too few entries. Found ${currentAgentData.length} entries, expected ${chunkLength}.`,
+                ErrorLevel.ERROR
             );
-        };
-
-        while (visData.length) {
-            const nSubPoints = visData[nSubPointsIndex];
-            const chunkLength = AGENT_OBJECT_KEYS.length + nSubPoints; // each array length is variable based on how many subpoints the agent has
-            if (visData.length < chunkLength) {
-                const attemptedMapping = AGENT_OBJECT_KEYS.map(
-                    (name, index) => `${name}: ${visData[index]}<br />`
-                );
-                // will be caught by controller.changeFile(...).catch()
-                throw new FrontEndError(
-                    "Your data is malformed, there are too few entries.",
-                    ErrorLevel.ERROR,
-                    `Example attempt to parse your data: <pre>${attemptedMapping.join(
-                        ""
-                    )}</pre>`
-                );
-            }
-
-            const agentSubSetArray = visData.splice(0, chunkLength); // cut off the array of 1 agent data from front of the array;
-            if (agentSubSetArray.length < AGENT_OBJECT_KEYS.length) {
-                const attemptedMapping = AGENT_OBJECT_KEYS.map(
-                    (name, index) => `${name}: ${agentSubSetArray[index]}<br />`
-                );
-                // will be caught by controller.changeFile(...).catch()
-                throw new FrontEndError(
-                    "Your data is malformed, there are less entries than expected for this agent. ",
-                    ErrorLevel.ERROR,
-                    `Example attempt to parse your data: <pre>${attemptedMapping.join(
-                        ""
-                    )}</pre>`
-                );
-            }
-
-            const agent = parseOneAgent(agentSubSetArray);
-            parsedAgentData.push(agent);
         }
 
-        const frameData: FrameData = {
-            time: frame.time,
-            frameNumber: frame.frameNumber,
-        };
+        agentCount++;
 
-        parsedAgentDataArray.push(parsedAgentData);
-        frameDataArray.push(frameData);
-    });
+        // Copy agent data
+        for (let i = 0; i < FRAME_DATA_SIZE; i++) {
+            view[writeIndex++] = frame.data[readIndex++];
+        }
 
-    return {
-        parsedAgentDataArray,
-        frameDataArray,
+        // Validate data integrity
+        if (--readIndex + nSubPoints > frame.data.length) {
+            throw new FrontEndError(
+                `Your data is malformed, there are too few entries. Found ${
+                    frame.data.length
+                }, expected ${readIndex + nSubPoints}.`,
+                ErrorLevel.ERROR
+            );
+        }
+
+        // Copy subpoints
+        for (let i = 0; i < nSubPoints; i++) {
+            view[writeIndex++] = frame.data[readIndex++];
+        }
```

Comment:
Optional suggestion: could try something like

```typescript
const subpoints = frame.data.slice(readIndex, readIndex + nSubPoints);
view.set(subpoints, writeIndex);
readIndex += nSubPoints;
writeIndex += nSubPoints;
```

The runtime may be able to turn that `set` call into something more efficient than a loop? If so, looks like there's a spot above to try it too?

Reply:
the numbers of subpoints are generally not huge. I'd benchmark this if you are really wondering if the suggestion is faster.  If both variables were typedarrays then you could just do a buffer copy for sure.

Reply:
Why not, let's go. We're typescript developers we shouldn't be writing loops ;)